Support `margin-trim: block inline` keyword combination

nt1m opened the pull request:
#### 598304c8891f893c09154ffc35588577338e50a6
<pre>
Support `margin-trim: block inline` keyword combination
<a href="https://bugs.webkit.org/show_bug.cgi?id=284978">https://bugs.webkit.org/show_bug.cgi?id=284978</a>
<a href="https://rdar.apple.com/141784069">rdar://141784069</a>

Reviewed by Sammy Gill.

Add parsing/serialization support for this combination as specified in: <a href="https://www.w3.org/TR/css-box-4/#margin-trim">https://www.w3.org/TR/css-box-4/#margin-trim</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-box/parsing/margin-trim-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/parsing/margin-trim-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/parsing/margin-trim-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/parsing/margin-trim.html:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Box.cpp:
(WebCore::CSSPropertyParserHelpers::consumeMarginTrim):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertMarginTrim):

Canonical link: <a href="https://commits.webkit.org/294190@main">https://commits.webkit.org/294190@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/31134996ca97b89ba9915a2c426396163c3c089c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100912 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20574 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10877 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106058 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51510 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102953 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20882 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29068 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76849 "Found 1 new test failure: imported/w3c/web-platform-tests/html/semantics/forms/constraints/infinite_backtracking.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33883 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103919 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16059 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91148 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57196 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15873 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9170 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50886 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85785 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9231 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108414 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28039 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20620 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85818 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28402 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87348 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85359 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30066 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7794 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22069 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16436 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27970 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33238 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27782 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31102 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29340 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sammygill commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -2174,7 +2174,14 @@ inline OptionSet<MarginTrimType> BuilderConverter::convertMarginTrim(BuilderStat
+            marginTrim.add({ MarginTrimType::InlineStart, MarginTrimType::InlineEnd });
+    }
+    if (!list->size())
+        return marginTrim;
```
Isn't this already handled a few lines above?

nt1m replied:
Ah I meant to write `if (!marginTrim->isEmpty())` :)

sammygill commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -2174,7 +2174,14 @@ inline OptionSet<MarginTrimType> BuilderConverter::convertMarginTrim(BuilderStat
     OptionSet<MarginTrimType> marginTrim;
-    ASSERT(list->size() <= 4);
+    for (auto& item : *list) {
+        if (item.valueID() == CSSValueBlock)
+            marginTrim.add({ MarginTrimType::BlockStart, MarginTrimType::BlockEnd });
+        if (item.valueID() == CSSValueInline)
+            marginTrim.add({ MarginTrimType::InlineStart, MarginTrimType::InlineEnd });
```
Can we just add these to the loop a few lines below?

nt1m replied:
You can't mix block/inline with the other keywords, so I decided to handle them separately in the converter

weinig commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Box.cpp:
```diff
@@ -45,20 +45,35 @@ RefPtr<CSSValue> consumeMarginTrim(CSSParserTokenRange& range, CSS::PropertyPars
 
     auto firstValue = range.peek().id();
-    if (firstValue == CSSValueBlock || firstValue == CSSValueInline || firstValue == CSSValueNone)
+    if (firstValue == CSSValueNone)
```
Please update the grammar entry in CSSProperties.json.

nt1m replied:
Done

nt1m committed:
Support `margin-trim: block inline` keyword combination
https://bugs.webkit.org/show_bug.cgi?id=284978
rdar://141784069

Reviewed by Sammy Gill.

Add parsing/serialization support for this combination as specified in: https://www.w3.org/TR/css-box-4/#margin-trim

* LayoutTests/imported/w3c/web-platform-tests/css/css-box/parsing/margin-trim-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/parsing/margin-trim-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/parsing/margin-trim-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/parsing/margin-trim.html:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Box.cpp:
(WebCore::CSSPropertyParserHelpers::consumeMarginTrim):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertMarginTrim):

Canonical link: https://commits.webkit.org/294190@main
